dplutils/pipeline/utils.py

```diff
@@ -16,3 +19,19 @@ def deque_extract(queue, condition):
             yield queue.pop()
         else:
             queue.rotate()
+
+
+def split_dataframe(df, max_rows=None, num_splits=None):
+    """Split dataframe by max number of rows or given splits.
+
+    If num_splits is provided, ignore max_rows and produce a set of num_splits
+    splits with each dataframe having approximately equal number of rows (one may be
+    different in size).
+
+    If max_rows provided, produce the smallest number of splits that ensure no
+    single split has more than max_rows
+    """
+    if num_splits is None:
+        num_splits = np.ceil(len(df) / max_rows)
+    chunks = np.linspace(0, num_splits, num=len(df), endpoint=False, dtype=np.int32)
+    return [chunk for _, chunk in df.groupby(chunks)]
```

Comment:
Curious to why you use group by here. If you're using it only for splitting the dataframe, then using `.iloc` is seems to be faster. However if you get something different from using `groupby` feel free to ignore. 

Here is what the function could look like using `.iloc`

```python 
import itertools
import numpy as np
import pandas as pd

def split_dataframe_iloc(df, max_rows=None, num_splits=None):
    """Split dataframe by max number of rows or given splits.
    If num_splits is provided, ignore max_rows and produce a set of num_splits
    splits with each dataframe having approximately equal number of rows (one may be
    different in size).
    If max_rows provided, produce the smallest number of splits that ensure no
    single split has more than max_rows
    """
    if num_splits is None:
        num_splits = np.ceil(len(df) / max_rows)
    
    split_idxs = np.linspace(0, len(df), int(np.ceil(len(df)/num_splits))+1, dtype=np.int32, endpoint=True)
    
    return [df.iloc[start:end] for start, end in itertools.pairwise(split_idxs)]
``` 

When I run %%timeit on my local I get the following:

![notebook](https://github.com/ssec-jhu/dplutils/assets/13540823/1e8d38c7-dd39-4f19-a01e-0fd4314e28ab)

Reply:
Hey @ryanhausen, thanks! One potential concern with using `iloc` is that it makes a view on the dataframe, and I think if the chunks, for whatever reason, end up in the same process as the original and the original is destined for different processing, this might cause issues. A copy would be more general here - groupby gives this. Or maybe a iloc + copy, or some other way?

Reply:
I timed the groupby vs the iloc with a copy and looks like groupby is quicker (for above case on my machine 152 us vs 286 us)

Reply:
@amitschang  

wrt to speed: `groupby` is also faster than `iloc` (with `.copy(deep=True)`) on my desktop, but not on my laptop. I am not sure what the difference is. I am also surprised that groupby is faster than iloc in at all, this feels counter intuitive.

Either way, if one way isn't obviously faster than another, then I don't think there is a strong reason to change what works. Thanks for indulging me!  I am going to look into this more now though haha.

Reply:
@amitschang  

wrt to speed: `groupby` is also faster than `iloc` (with `.copy(deep=True)`) on my desktop, but not on my laptop. I am not sure what the difference is. I am also surprised that groupby is faster than iloc at all, this feels counter intuitive.

Either way, if one way isn't obviously faster than another, then I don't think there is a strong reason to change works. Thanks for induldging!

Reply:
...maybe pandas is using cython for its splits in groupby and we're working at the python level.

I need to let this go. Good to know though!

Reply:
Thanks Ryan, I'll go with whats here for the above reasons, but thanks for the suggestion!